### ISSUE TITLE
Multi-core support

### ISSUE THREAD

**Issue by trifud** [samtools/samtools#348]
It would be nice if multi-cpu and multi-core support could be added. This will grealy reduce processing time for large files on more capable machines.

**Comment by tseemann** [comment]
Many of the samtools sub-tools support the ```-@ INT``` option which is the number of threads to use. Supported by ```view``` and ```sort``` for example.

**Comment by jkbonfield** [comment]
More commands will soon be acquiring multithreaded BAM encoding and multithreaded CRAM encoding and decoding via https://github.com/samtools/samtools/pull/395 once it is finalised.

However it's a little bit cryptic - basically it means that a tool can set the nthreads option manually irrespective of whether a -@ option exists. Ideally we'd make -@ universal to obscure the internals a bit more.

**Comment by tseemann** [comment]
Thanks for the clarification James.

**Comment by jmarshall** [comment]
This will be at least partially addressed by PR #395.

**Comment by cchng** [comment]
Hi, just double checking if ``-@`` is supported with the ``depth`` command in Version: 1.3.1, or if it could somehow be used with regions to get depth of coverage? Thanks!

**Comment by jkbonfield** [comment]
Making -@ universal is due to the easy win of allowing all commands to uncompress blocks of data in parallel (for both BAM and CRAM).  It doesn't mean however that each algorithm using the data can automatically run in parallel (as it's limited to just the decoding portion).  It may be possible to extend algorithms to do more work in parallel, but that's another more complex task.

In this case it means we haven't got a version of `depth` that operates over multiple regions at the same time.

**Comment by pieterlukasse** [comment]
any updates on this one?

**Comment by jkbonfield** [comment]
Multi-core support for decoding and encoding of file formats is now universal.  Thus more or less every sub-command now does this.

I'm closing this issue as it's not specific enough.  Pretty much all samtools sub-commands do have multi-core support and have done for ages. What we haven't done is added "chunking" type methods to the core algorithms like depth. mpileup, etc to permit further multi-threading capabilities.  However those are separate specific issues related to specific algorithms and not well tracked by a blanket request as this issue.

**Comment by pieterlukasse** [comment]
@jkbonfield thanks for the update. In my case, I was referring specifically to the `index` sub-command. Should I log this as a new ticket?

**Comment by jkbonfield** [comment]
Samtools index is already multi-threaded.  What problem do you have with it?

**Comment by pieterlukasse** [comment]
Hi @jkbonfield, when I run it, I only see one of the cores busy. Apparently `samtools index` is not multi-threaded by default. Which parameter should I set? Thanks!

**Comment by jkbonfield** [comment]
None of the tools use multiple cores by default.  Use the `-@` option.  From the help:

```
Usage: samtools index [-bc] [-m INT] <in.bam> [out.index]
Options:
  -b       Generate BAI-format index for BAM files [default]
  -c       Generate CSI-format index for BAM files
  -m INT   Set minimum interval size for CSI indices to 2^INT [14]
  -@ INT   Sets the number of threads [none]
```

**Comment by pieterlukasse** [comment]
thanks @jkbonfield ! Somehow this information seems to be missing in the manpage. I'll try it out.

**Comment by tseemann** [comment]
The word `thread` is all over the man page?

```
% man samtools

                 Finally, the -@ option can be used to allocate additional threads to be used  for  com‐
                 pression, and the -?  option requests a long help message.
```

**Comment by jmarshall** [comment]
@tseemann: It's mentioned specifically for **view** and **sort**, but not for other commands. On **develop** the man pages have been split out per command, so for example [samtools-index.1](https://github.com/samtools/samtools/blob/develop/doc/samtools-index.1) mentions all of **index**'s options except `-@`. So there's room for improvement in synchronising man pages and usage displays.

**Comment by jkbonfield** [comment]
Thanks, that's a good point.  I hadn't noticed we missed some of the more global options in the split man pages.  I generated those mainly by simply excising out the sections in the monolithic man, so I guess it was also missing there.  Most likely I forgot to add them when I added it as a mostly-universal option.

Something for the to-do list.

**Comment by apallav** [comment]
Could mpileup be used -@ int as well?

**Comment by jkbonfield** [comment]
The main mpileup needs breaking up into regions so it can be multithreaded properly.  If I recall I tried multi-theading the BAM and CRAM decoders for mpileup and the different was almost nil as it spends nearly all its time in mpileup itself rather than file decoding.  It's one of the many things in my mental to-do list.

**Comment by apallav** [comment]
Thanks!
